File: plug-in/PDFArchiveFile/VB.NET/ChooseBackupFolder.vb
Imports System.Windows.Forms
Imports Microsoft.WindowsAPICodePack.Dialogs

Public Class ChooseBackupFolder
    ' Event handler for the "Browse Folder" button click
    Private Sub btnBrowseFolder_Click(sender As Object, e As EventArgs) Handles btnBrowseFolder.Click
        ' Open a folder picker dialog to allow the user to select a folder
        Dim dialog As CommonOpenFileDialog = New CommonOpenFileDialog()
        dialog.IsFolderPicker = True ' Ensure the dialog only allows folder selection
        If dialog.ShowDialog() = CommonFileDialogResult.Ok Then
            ' Set the selected folder path to the text box
            txtBackupFolder.Text = dialog.FileName
        End If
    End Sub

    ' Event handler for the "OK" button click
    Private Sub btnOK_Click(sender As Object, e As EventArgs) Handles btnOK.Click
        ' Save the selected folder path to the application settings
        SaveSetting(PDFArchiveFile.WIN2PDF_COMPANY, PDFArchiveFile.WIN2PDF_PRODUCT, PDFArchiveFile.ARCHIVE_FOLDER_SETTING, txtBackupFolder.Text)
        ' Close the form
        Me.Close()
    End Sub

    ' Event handler for the "Cancel" button click
    Private Sub btnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        ' Close the form without saving any changes
        Me.Close()
    End Sub

    ' Method to center the form on the screen or relative to a parent form
    ' Parameters:
    ' - frm: The form to be centered
    ' - parent: (Optional) The parent form relative to which the form will be centered
    Private Sub CenterForm(ByVal frm As Form, Optional ByVal parent As Form = Nothing)
        Dim r As Drawing.Rectangle
        If parent IsNot Nothing Then
            ' Get the rectangle of the parent form's client area
            r = parent.RectangleToScreen(parent.ClientRectangle)
        Else
            ' Get the working area of the screen where the form is located
            r = Screen.FromPoint(frm.Location).WorkingArea
        End If

        ' Calculate the new position to center the form
        Dim x = r.Left + (r.Width - frm.Width) \ 2
        Dim y = r.Top + (r.Height - frm.Height) \ 2
        frm.Location = New Drawing.Point(x, y)
    End Sub

    ' Event handler for the form's Load event
    Private Sub ChooseBackupFolder_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        ' Load the previously saved backup folder path from application settings
        txtBackupFolder.Text = Interaction.GetSetting(PDFArchiveFile.WIN2PDF_COMPANY, PDFArchiveFile.WIN2PDF_PRODUCT, PDFArchiveFile.ARCHIVE_FOLDER_SETTING, My.Computer.FileSystem.SpecialDirectories.MyDocuments + "\backup")
        ' Center the form on the screen
        CenterForm(Me)
    End Sub

End Class

File: plug-in/PDFArchiveFile/C#/ChooseBackupFolder.cs
using System;
using System.Windows.Forms;
using Microsoft.VisualBasic;

namespace PDFChooseBackupFolder
{
    public partial class ChooseBackupFolder : Form
    {
        // Constructor to initialize the form
        public ChooseBackupFolder()
        {
            InitializeComponent();
        }

        /// <summary>
        /// Centers the given form on the screen or relative to a parent form.
        /// </summary>
        /// <param name="frm">The form to center.</param>
        /// <param name="parent">The parent form to center relative to (optional).</param>
        private void CenterForm(Form frm, Form parent = null/* TODO Change to default(_) if this is not a reference type */)
        {
            System.Drawing.Rectangle r;

            // Determine the area to center the form within
            if (parent != null)
                r = parent.RectangleToScreen(parent.ClientRectangle);
            else
                r = Screen.FromPoint(frm.Location).WorkingArea;

            // Calculate the centered position
            var x = r.Left + (r.Width - frm.Width) / 2;
            var y = r.Top + (r.Height - frm.Height) / 2;

            // Set the form's location to the calculated position
            frm.Location = new System.Drawing.Point(x, y);
        }

        /// <summary>
        /// Opens a folder browser dialog to allow the user to select a backup folder.
        /// </summary>
        private void btnBrowseFolder_Click(object sender, EventArgs e)
        {
            FolderBrowserDialog folderBrowserDialog1 = new FolderBrowserDialog();

            // Show the folder browser dialog and set the selected path to the text box
            if (folderBrowserDialog1.ShowDialog() == DialogResult.OK)
            {
                txtBackupFolder.Text = folderBrowserDialog1.SelectedPath;
            }
        }

        /// <summary>
        /// Closes the form when the Cancel button is clicked.
        /// </summary>
        private void btnCancel_Click(object sender, EventArgs e)
        {
            this.Close();
        }

        /// <summary>
        /// Saves the selected backup folder path to the application settings and closes the form.
        /// </summary>
        private void btnOK_Click(object sender, EventArgs e)
        {
            // Save the selected folder path to the application settings
            Interaction.SaveSetting(PDFArchiveFile.WIN2PDF_COMPANY, PDFArchiveFile.WIN2PDF_PRODUCT, PDFArchiveFile.ARCHIVE_FOLDER_SETTING, txtBackupFolder.Text);

            // Close the form
            this.Close();
        }

        /// <summary>
        /// Loads the form, initializes the backup folder path, and centers the form.
        /// </summary>
        private void ChooseBackupFolder_Load(object sender, EventArgs e)
        {
            // Retrieve the saved backup folder path or use a default path
            txtBackupFolder.Text = Interaction.GetSetting(PDFArchiveFile.WIN2PDF_COMPANY, PDFArchiveFile.WIN2PDF_PRODUCT, PDFArchiveFile.ARCHIVE_FOLDER_SETTING, Environment.SpecialFolder.MyDocuments + @"\backup");

            // Center the form on the screen
            CenterForm(this);
        }
    }
}


//namespace PDFDuplicateFile
//{
//    public partial class ChooseBackupFolder : Form
//    {
//        public ChooseBackupFolder()
//        {
//            InitializeComponent();
//        }
//    }
//}

